core/blockchain_bor_test.go

```diff
@@ -0,0 +1,123 @@
+package core
+
+import (
+	"math/big"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/consensus/ethash"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+func TestChain2HeadEvent(t *testing.T) {
+	var (
+		db      = rawdb.NewMemoryDatabase()
+		key1, _ = crypto.HexToECDSA("b71c71a67e1177ad4e901695e1b4b9ee17ae16c6668d313eac2f96dbcda3f291")
+		addr1   = crypto.PubkeyToAddress(key1.PublicKey)
+		gspec   = &Genesis{
+			Config: params.TestChainConfig,
+			Alloc:  GenesisAlloc{addr1: {Balance: big.NewInt(10000000000000000)}},
+		}
+		genesis = gspec.MustCommit(db)
+		signer  = types.LatestSigner(gspec.Config)
+	)
+
+	blockchain, _ := NewBlockChain(db, nil, gspec.Config, ethash.NewFaker(), vm.Config{}, nil, nil)
+	defer blockchain.Stop()
+
+	chain2HeadCh := make(chan Chain2HeadEvent, 64)
+	blockchain.SubscribeChain2HeadEvent(chain2HeadCh)
+
+	chain, _ := GenerateChain(gspec.Config, genesis, ethash.NewFaker(), db, 3, func(i int, gen *BlockGen) {})
+	if _, err := blockchain.InsertChain(chain); err != nil {
+		t.Fatalf("failed to insert chain: %v", err)
+	}
+
+	replacementBlocks, _ := GenerateChain(gspec.Config, genesis, ethash.NewFaker(), db, 4, func(i int, gen *BlockGen) {
+		tx, err := types.SignTx(types.NewContractCreation(gen.TxNonce(addr1), new(big.Int), 1000000, gen.header.BaseFee, nil), signer, key1)
+		if i == 2 {
+			gen.OffsetTime(-9)
+		}
+		if err != nil {
+			t.Fatalf("failed to create tx: %v", err)
+		}
+		gen.AddTx(tx)
+	})
+
+	if _, err := blockchain.InsertChain(replacementBlocks); err != nil {
+		t.Fatalf("failed to insert chain: %v", err)
+	}
+
+	i := 0
+	readEvent := func() *Chain2HeadEvent {
+		select {
+		case ev := <-chain2HeadCh:
+			i++
+			return &ev
+		case <-time.After(2 * time.Second):
+			t.Fatal("timeout")
+		}
+		return nil
+	}
+
+	// head event
+	event1 := readEvent()
```

Comment:
I would recommend you yet another iteration for this tests. Instead of having to manually write down every comparison, you can send to readEvent a struct of what you expect to be seen on the event. For example:

```
type eventTest struct {
    Type Chain2HeadType
    Added []common.Hash
    Removed []common.Hash
}

readEvent := func(expect *eventTest) {
    var ev *Chain2HeadEvent
    select {
        case ev = <-chain2HeadCh:
        case <-time.After(...:
            t.Fatal("timeout")
    }
    // compare ev with expect ...
    // compare Type
    // compare Added hashes
    // compare Removed hashes
}

// event 1
readEvent(&eventTest{
    Type: Chain2HeadCanonicalEvent,
    Added: []common.Hash{
        chain[2].Hash(),
    },
})

// event 2..
// event 3..
```